bayes_opt1/util.py

```diff
@@ -0,0 +1,547 @@
+            x_newobs = x_obs
+            y_newobs = y_obs
+        else:
+            y=np.random.normal(loc=mean, scale=std)
+            x_newobs=np.concatenate((x_obs, i.reshape(1,-1)), axis = 0)
+            y_newobs=np.append(y_obs,y)
+        mu=posterior(gp, x_newobs, y_newobs, grid)
+        return (max(mu),np.argmax(mu))
+    
+    def mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,i):
+        '''
+        see Algorithm 4:
+        In our setting, we apply this by letting x_star_hat be the x^prime maximizing µ_{n+1}(x^prime; x, µ_n(x)+σ_n(x)Z), 
+        and then calculating the gradient of µ_{n+1}(x_star_hat; x, µ_n(x)+σ_n(x)Z) with respect to x 
+        while holding x_star_hat fixed.
+        '''
+        y=np.random.normal(loc=mean, scale=std)
+        x_newobs=np.concatenate((x_obs, i.reshape(1,-1)), axis = 0)
+        y_newobs=np.append(y_obs,y)
+        mu=posterior(gp, x_newobs, y_newobs, x_star_hat)
+        return np.asscalar(mu)
+    
+    p = optimizer._space.bounds.shape[0] # size of state space (e.g. if we are working on R^3, n = 3)
+    init = [np.zeros(n_grid)] * p # list of p vectors, each of lenght n_grid   
+    x_obs_temp = np.array([[res["params"]] for res in optimizer.res]) 
+    y_obs = np.array([res["target"] for res in optimizer.res])
+    # x_obs needs to be np.array of size (n,p)
+    # y_obs needs to be np.array of size (n,)
+    # see https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html
+    x_obs = x_obs_to_array(x_obs_temp, p)
+
+    for i in range(p): 
+        init[i] = np.linspace(optimizer._space.bounds[i,0], optimizer._space.bounds[i,1], n_grid) # works \forall p
+
+    grid = np.meshgrid(*init) # list with p matrices n_grid x n_grid
+    for g in range(len(grid)):
+        grid[g] = grid[g].reshape(-1,1)
+    grid = np.stack(grid, axis = -1)
+    grid=grid[:,0,:] # array of shape (n_grid*n_grid, p): each row is composed of one element from init[0] and one element from init[1]
+    
+    temp = find_mean_max(x_obs,y_obs,gp,grid)
+    mean_nstar = temp[0]
+
+    grad_temp = []
+    mean, std= gp.predict(x.reshape(1,-1), return_std=True)
+    for j in range(J):
+        print("j = ",j+1)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            temp = find_mean_max(x_obs,y_obs,gp,grid,mean,std,x)
+            mean_n1star = temp[0]
+            x_star_hat = grid[temp[1]].reshape(1,-1)
+            def func(w):
+                return mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,w)
+            
+            grad_temp.append(nd.Gradient(func)(x))
+    grad = np.zeros(p)
+    for i in grad_temp:
+        grad = grad + i
+    grad = grad/J
+    return grad
+
+
+
+# Implementation algorithm 3 (Frazier): Knowledge gradient 
+def minimize_kg_sgd(R, T, a, p, pbounds, optimizer, gp):
+
+    alpha = np.zeros(T)
+    KG = np.zeros(R)
+    temp = our_lhs(p,R,pbounds)
```

Comment:
Cambierei nome a `temp` (in realtà anche altre cose che avete chiamato "temp", ma questa per qualche ragione mi ha colpito di più...)